### ISSUE TITLE
Will it support DPO(direct preference optimization) or other RLHF methods?

### ISSUE THREAD

**Issue by yechenzhi** [pytorch/torchtune#621]

**Comment by RdoubleA** [comment]
Thanks for raising this - no plans at the moment, we will reassess this after release and see if it's a strong community need

**Comment by kartikayk** [comment]
@yechenzhi what's the use case you have in mind here? And are there any tools that you use today for DPO and other reward modeling?

BTW is this something you'd be interested in contributing?

**Comment by yechenzhi** [comment]
I am currently exploring various alignment techniques for tuning LLMs using [TRL](https://github.com/huggingface/trl). I've noticed that Torchtune appears to run more quickly, although I haven't conducted a comprehensive comparison of the experimental settings. Personally, I find Torchtune's coding style more to my liking, which has led me to consider why not adopt Torchtune for our experiments.

Furthermore, if there's a need to integrate DPO or otherRLHF methods into Torchtune, I would be eager to give it a try.

**Comment by kartikayk** [comment]
@yechenzhi glad to hear TorchTune's been helpful!

Yes! We'd love to integrate DPO to the library. I think this would be a pretty awesome contribution. Do you want to take a stab at putting together a quick RFC/design for this? just open an issue with your thoughts and some pseudo-code and we can start a discussion. I think this shouldn't take a lot of discussion to be added to the library. 

A few things to keep in mind:

- We try to minimize third-party dependencies as much as possible and want to write PyTorch as much as possible
- IMO, this would be a recipe that you add to ```recipes/```. Take a look at the [recipe interface](https://github.com/pytorch/torchtune/blob/main/torchtune/recipe_interfaces.py) and see if this helps provide more information about how we think about recipes

Let me know if you have any questions!

cc: @vmoens

**Comment by yechenzhi** [comment]
@kartikayk Hello, I wrote a brief [RFC](https://github.com/pytorch/torchtune/issues/623#issue-2217054688) that includes preliminary pseudo-code, which is more of a conceptual framework at this stage. Please let me know if further pseudo-code is required.

**Comment by kartikayk** [comment]
Thanks so much for the contribution @yechenzhi! Sorry for the delay in reviewing the multi-GPU training recipe. We got caught up with the launch, but this will be something we look at ASAP. I'm closing this issue - thanks again for th great work!
